src/e3/testsuite/driver/diff.py

```diff
@@ -0,0 +1,314 @@
+        """
+        :param str pattern: Pattern (regular expression) to replace.
+        :param str replacement: Replacement to use for the substitution.
+        """
+        self.regexp = re.compile(pattern)
+        self.replacement = replacement
+
+    def refine(self, output):
+        return self.regexp.sub(self.replacement, output)
+
+
+def ReplacePath(path, replacement=""):
+    """Return an output refiner to replace the given path."""
+
+    # TODO: the path processings below were mostly copied from gnatpython.
+    # The exact intend behind them is unknown: we should investigate
+    # removing them at some point, and if they are really needed, document
+    # here why they are.
+
+    def escape(s):
+        return s.replace("\\", "\\\\")
+
+    return RefiningChain([
+        Substitute(substring, replacement)
+        for substring in [escape(os.path.abspath(path)),
+                          escape(unixpath(path)),
+                          escape(path)]
+    ])
+
+
+class DiffTestDriver(ClassicTestDriver):
+    """Test driver to compute test output against a baseline."""
+
+    @property
+    def baseline_file(self):
+        """Return the test output baseline file.
+
+        :return: The name of the text file (relative to test directories) that
+            contains the expected test output and whether the baseline is a
+            regexp.
+        :rtype: (str, bool)
+        """
+        filename = self.test_env.get("baseline_file", "test.out")
+        is_regexp = self.test_env.get("baseline_regexp", False)
+        return (filename, is_regexp)
+
+    @property
+    def baseline(self):
+        """Return the test output baseline.
+
+        Subclasses can override this method if they want to provide a baseline
+        that does not come from a file, short-circuiting the baseline_file
+        property.
+
+        :return: The baseline absolute filename (if any), the baseline content,
+            as a string or as a bytes string (depending on the default
+            encoding), and whether the baseline is a regexp. The baseline
+            filename is used to rewrite test output: leave it to None if
+            rewriting does not make sense.
+        :rtype: (str|None, str|bytes, bool)
+        """
+        filename, is_regexp = self.baseline_file
+        filename = self.test_dir(filename)
+        is_binary = self.default_encoding == "binary"
+        mode = "rb" if is_binary else "rt"
+
+        try:
+            with open(filename, mode) as f:
+                baseline = f.read()
+            if not is_binary and sys.version_info.major == 2:  # py2-only
+                baseline = baseline.decode(self.default_encoding)
+        except Exception as exc:
+            raise TestAbortWithError(
+                "cannot read baseline file ({}: {})".format(
+                    type(exc).__name__, exc
+                )
+            )
+        return (filename, baseline, is_regexp)
+
+    output_refiners = []
```

Comment:
isn't it dangerous to have class variable here?

Reply:
As discussed live: there are supposed to be constants. I’ll turn them into properties to make that explicit, although users will be able to override them with class attributes (and that’s fine).